api/envoy/extensions/upstreams/http/v3/http_protocol_options.proto

```diff
@@ -0,0 +1,94 @@
+syntax = "proto3";
+
+package envoy.extensions.upstreams.http.v3;
+
+import "envoy/config/core/v3/protocol.proto";
+
+import "udpa/annotations/status.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.upstreams.http.v3";
+option java_outer_classname = "HttpProtocolOptionsProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: HTTP Protocol Options]
+// [#extension: envoy.upstreams.http.http_protocol_options]
+
+// HttpProtocolOptions specifies Http upstream protocol options. This object
+// is used in
+// :ref:`typed_extension_protocol_options<envoy_api_field_config.cluster.v3.Cluster.typed_extension_protocol_options>`,
+// keyed by the name `envoy.extensions.upstreams.http.v3.HttpProtocolOptions`.
+//
+// This controls what protocol(s) should be used for upstream and how said protocol(s) are configured.
+//
+// This replaces the prior pattern of explicit protocol configuration directly
+// in the cluster. So a configuration like this, explicitly configuring the use of HTTP/2 upstream:
+//
+// .. code::
+//
+//   clusters:
+//     - name: some_service
+//       connect_timeout: 5s
+//       upstream_http_protocol_options:
+//         auto_sni: true
+//       common_http_protocol_options:
+//         idle_timeout: 1s
+//       http2_protocol_options:
+//         max_concurrent_streams: 100
+//        .... [further cluster config]
+//
+// Would now look like this:
+//
+// .. code::
+//
+//   clusters:
+//     - name: some_service
+//       connect_timeout: 5s
+//       typed_extension_protocol_options:
+//        envoy.extensions.upstreams.http.v3.HttpProtocolOptions:
+//           "@type": type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions
```

Comment:
I wonder if, given that we have this type URL anyway, the key is actually purely advisory at this point anyway. I think this would require some followup changes unrelated to this PR. @kyessenov @lizan thoughts? This seems a good direction to head.

Reply:
Agree that the name is a stutter. It looks like there's some implicit protocol match though, so it's also reasonable to have  an a concrete proto with field for every protocol. I think it depends what is the intent on the extensions here. Are there are other HTTP options planned? Are there opaque options for future protocols?

Reply:
there are other options for other protocols, (redis for example)
agree if we're ditching key, the map isn't necessary, but I'm going to ask someone else shave that yak if it's OK :-)

Reply:
Yeah, I think it becomes a list of typed extensions in the future. @alyssawilk could you leave some trail of this thinking in comment in a `[#comment:` block? I agree it's a distinct piece of work, just want to make sure we don't forget this.